Workshop Doctrine Xavi

xserrat opened the pull request:
Bones @obokaman-com ,

Afegeixo PR de la part de Doctrine del Workshop de Symfony. A la iteració que hi ha ara estan els mapejos amb yaml amb els models del domini. Fins al moment d'afegir Skills tot anava bé, però al afegir `Skill` i `SkillCollection` arriba el primer problema: Doctrine espera un `ArrayCollection` i com a entitat de domini hi tinc un `SkillCollection`....

Vaig a per la segona interació fent-ho amb entities de doctrine per un costat i entitats de domini per l'altre!

Commit:
xavi Remove default AppBundle

Commit:
xavi Add initial UserBundle

Commit:
xavi Add users table to sqlite

Commit:
xavi Add User entity and some value objects

xserrat committed:
xavi Refactor user id in order to use a varchar type

xserrat committed:
xavi Change column id type to varchar instead of integer

xserrat committed:
xavi Move entities and value objects from Domain parent folder to the new Model folder located into the Domain folder

xserrat committed:
xavi Add the add user use case

- Add the application service and the request.
- Add UserRepository contract to the domain layer.
- Partial implementation of Dbal UserRepository.
- Define services to the services.yml

xserrat committed:
xavi Add register template and UserController with a register form to add users

xserrat committed:
xavi Add register form and list view. Add initial use cases in order to delete, get user by id and update it

Commit:
xavi Add complete use case to delete a user from list template

Commit:
xavi Add button to register a user and to back to the user's list

Commit:
xavi Add CacheBundle to manage userRepository output

- Add domain contract for Cache implementations and implement FilesystemCache
- Add UserCacheRepository that decorates the initial UserRepository

- Cache bundle needs a good config to work properly!

Commit:
xavi Remove unused Controller, view and remove old repository

Commit:
xavi Remove routing.yml from CacheBundle and implement FilesystemCache::set method

Commit:
xavi Add gitignore with cache file

Commit:
xavi Change cache file path to the /var/cache/dev

Commit:
Add first approach using Events

xserrat committed:
xavi Second approach to throw events with a DomainEventDispatcher

Commit:
xavi Refactor DDD structure and move logic outside. Also remove unused Symfony Configuration class from the two bundles

Commit:
xavi Move UserCacheRepository from Domain/Infrastructure to Infrastructure

Commit:
xavi: Add EventSubscriber to invalidate user cache when adding a new user or deleting an existent

Commit:
Remove unused DomainEventSubscriber contract

Commit:
xavi: Move DependencyInjection logic to the DDD structure as a 'Infrastructure/DependencyInjection/Symfony'

Commit:
xavi: Decouple ReadModel application services from WriteModel

Commit:
xavi: Add get user by id use case

Commit:
xavi: Add update use case with the corresponding UserUpdated event for cache invalidation.

Commit:
xavi: Add doctrine configuration in order to mapping Domain\Model\User as a doctrine entity using yaml

Add embedded fields in order to obtain User object with UserId and Email instances inside.

Commit:
xavi: Refactor repository dbal and orm in order to force return a User domain model.

Also, add marshaller in order to transform User model to array used in twig template

Commit:
xavi: Force return UserCollection for UserRepository::findAll()

Also, add marshmaller to transform UserCollection into plain array

Commit:
xavi: Fix Filesystem cache serializing the whole object returned by repository

Commit:
xavi: Rename namespace for UserUpdated event from Workshop\UserManager... to UserManager\...

Commit:
xavi: Refactor event subscriber management applying use cases for each user cache invalidation

Commit:
xavi: Fix value passed to the UserDeleted event

Commit:
xavi: Implement DomainEventRecorder and DispatchEvents in order to dispatch all the events through a new layer DispatchEvents and record events from User aggregate

Also, fix some namespaces, add Username value object and store to the UserAdded event all user data in order to be able to store events with all data changes

Commit:
xavi: Fix Username mapping namespace and move EventSubscribers to the Application namespace instead of Infrastructure

Commit:
xavi: Some fixes and add three events in order to invalidate cache when name, surname or email has been updated

Commit:
xavi: Remove ValueObject directory in order to put all objects to the root User directory. Add Skill and SkillCollection classes and UserHasAddedANewSkill event when someone call the User::addSkill() method.

Commit:
xavi: Add user-skill mapping doctrine orm yml and update find queries to get list of users and an specific user by id

Commit:
xavi: Change to DoctrineORM instead of DoctrineDbal and see a PersistentCollection returned as a list of skills.

Commit:
xavi: Add first approach to persisting a mapped domain entity using yml and having a 1-N association inside a User.

This approach doesn't work properly because the skill collection returned by doctrine is a PersistentCollection, so it's impossible to persist a Domain entity using in our domain a SkillCollection and at the same time a PersitentCollection in order to be recognized by doctrine.

Commit:
xavi-doctrine: Refactor doctrine orm creating Doctrine Entities and perform the mapping between Doctrine entity and Domain entity in the UserRepository

obokaman-com reviewed (changes requested):
Em salta un error a l'hora d'accedir a la instància: 

Attempted to load class "GetAllUsersUseCase" from namespace "UserManager\ReadModel\Application\Service\User\GetAll".
Did you forget a "use" statement for another namespace?

Commit:
Revert "xavi-doctrine: Refactor doctrine orm creating Doctrine Entities and perform the mapping between Doctrine entity and Domain entity in the UserRepository"

Revert changes in order to use the orm mapping with domain entities

This reverts commit 8d18fab523d1918876903ecc3a0566979ab53a81.

Commit:
xavi-doctrine: Couple to the Doctrine's ArrayCollection inside User aggregate at the skills collection in order to allow Doctrine entities management with a collection.

In order to use Domain Collection and using Doctrine ORM, a way to fetch and persist entities with Doctrine's Entity Manager is coupling our Domain User aggregate with the ArrayCollection, but only on terms of management inside the aggregate (as a class attribute), but for outside usages we use the SkillCollection class.

Commit:
xavi-doctrine: Remove UseCase sufix from Application Services.

Commit:
xavi-doctrine: Remove unused return value on delete user

Commit:
xavi-doctrine: Create form type User and Skill

Commit:
xavi-doctrine: Add skills as a prototype in the User form in order to add dinamically skills. Although, the skill updating not works yet
